lib/ecto/embedded.ex

```diff
@@ -159,13 +159,17 @@ defmodule Ecto.Embedded do
   end
 
   defp autogenerate(changes, action, schema) do
-    Enum.reduce schema.__schema__(action_to_auto(action)), changes, fn
-      {k, {mod, fun, args}}, acc ->
-        case Map.fetch(acc, k) do
-          {:ok, _} -> acc
-          :error   -> Map.put(acc, k, apply(mod, fun, args))
-        end
-    end
+    autogen_fields = schema.__schema__(action_to_auto(action))
+
+    Enum.reduce(autogen_fields, changes, fn {fields, {method, fun, args}}, acc ->
+      generated = apply(method, fun, args)
```

Comment:
We should only `apply` if at least one field is not set. So I think we need to change things a bit. Something like:
```elixir
Enum.reduce(autogen_fields, changes, fn {fields, {method, fun, args}}, acc ->
  case Enum.reject(fields, Map.has_key?(acc, &1)) do
    [] ->
      acc

    fields ->
      generated = apply(method, fun, args)
      Enum.reduce(fields, acc, &Map.put(&2, &1, generated))
  end
end)
```